src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java

```diff
@@ -21,520 +21,550 @@
+         );
+         worker.setSkillModifier(2 * worker.getCitizenData().getEndurance() + worker.getCitizenData().getCharisma());
+         worker.setCanPickUpLoot(true);
+     }
+
+     /**
+      * Redirects the fisherman to his building.
+      *
+      * @return the next state.
+      */
+     private AIState startWorkingAtOwnBuilding()
+     {
+         if (walkToBuilding())
+         {
+             return getState();
+         }
+         return PREPARING;
+     }
+
+     /**
+      * Prepares the farmer for farming.
+      * Also requests the tools and checks if the farmer has sufficient fields.
+      *
+      * @return the next AIState
+      */
+     @NotNull
+     private AIState prepareForFarming()
+     {
+         @Nullable BuildingFarmer building = getOwnBuilding();
+
+         if (building == null || building.getBuildingLevel() < 1)
+         {
+             return AIState.PREPARING;
+         }
+
+         building.syncWithColony(world);
+
+         if (building.getFarmerFields().size() < getOwnBuilding().getBuildingLevel() && !building.assignManually())
+         {
+             searchAndAddFields();
+         }
+
+         if (building.hasNoFields())
+         {
+             chatSpamFilter.talkWithoutSpam("entity.farmer.noFreeFields");
+             return AIState.PREPARING;
+         }
+
+         //If the farmer has no currentField and there is no field which needs work, check fields.
+         if (building.getCurrentField() == null && building.getFieldToWorkOn() == null)
+         {
+             building.resetFields();
+             return AIState.IDLE;
+         }
+
+         @Nullable Field currentField = building.getCurrentField();
+
+         if (currentField.needsWork())
+         {
+             switch (currentField.getFieldStage())
+             {
+                 case EMPTY:
+                     if (!checkForHoe())
+                     {
+                         return walkToBlock(currentField.getLocation()) ? AIState.PREPARING : AIState.FARMER_HOE;
+                     }
+                     break;
+                 case HOED:
+                     if (canGoPlanting(currentField, building))
+                     {
+                         return walkToBlock(currentField.getLocation()) ? AIState.PREPARING : AIState.FARMER_PLANT;
+                     }
+                     break;
+                 case PLANTED:
+                     return walkToBlock(currentField.getLocation()) ? AIState.PREPARING : AIState.FARMER_HARVEST;
+                 default:
+                     break;
+             }
+         }
+         else
+         {
+             getOwnBuilding().setCurrentField(null);
+         }
+         return AIState.PREPARING;
+     }
+
+     /**
+      * Searches and adds a field that has not been taken yet for the farmer and then adds it to the list.
+      */
+     private void searchAndAddFields()
+     {
+         Colony colony = worker.getColony();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'colony' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)